### DISCUSSION TITLE
Best practices for using MaterialBuilder

### DISCUSSION THREAD

**Discussion by rawnsley** [google/filament#5819] [Q&A] [answered]
I've got some questions around MaterialBuilder that I would appreciate some insight on:

1. Do calls to `MaterialBuilder.build()` and `Material.build()` need to be made from the filament thread? (typically the main thread, but actually not in my application). They are in the examples, but I seem to be getting away with calling them on random threads.
2. Building materials with `MaterialBuilder.Optimization.PERFORMANCE` takes much longer than `MaterialBuilder.Optimization.NONE`. What is the actual benefit?
3. Material building can take a long time and hangs the filament thread by default. Presumably we should provide an alternative `JobSystem` to prevent this, but am I right in thinking this isn't exposed through the Java API yet?

**Comment by romainguy** [comment, marked as answer]
1. `Material.build()` needs to be called from the Filament thread, `MaterialBuilder.build()` doesn't need to (but call it from a single thread).

2. `PERFORMANCE` compiles the shaders to SPIR-V (IR format), runs a full optimizer, then regenerates the GLSL shaders. This greatly reduces the size of the shaders, tends to side-step driver bugs, and overall improves performance. But as you noticed, it's a slow step, which is why it's better done by `matc` offline.

3. Which material building are you talking about? `Material.build()` shouldn't do too much and needs to run on Filament's thread. If you are talking about `MaterialBuilder`, then yes, it's a heavy process (generating all the material variants, parsing for validity, etc.). Using `filamat-lite` helps as it skips optimizations and runs lighter weight versions of the some of the validation.

**Reply by rawnsley** [reply]
Thanks for this info.

I can confirm that calling `Material.build()` is quick so it doesn't seem to matter that it has to be on the filament thread.

`MaterialBuilder.build()` can be slow in `PERFORMANCE` mode, but if it doesn't have to be on the filament thread then it doesn't matter to me. I cache the result anyway so it's only ever called once per material combination. I've now added an extension function with a synchronization wrapper for this method to prevent multiple thread calls.

Regarding filament-lite: I've been using this happily for years now because it helped me keep my whole APK below 10MB, but I'm going to experiment with `PERFORMANCE` mode for a while. Even an increase of a few FPS would be worth the trade off.
